agent/cert-monitor/cert_monitor.go

```diff
@@ -0,0 +1,468 @@
+	}
+	if err := m.cache.Prepopulate(cachetype.ConnectCALeafName, certRes, leafReq.Datacenter, leafReq.Token, leafReq.Key()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Start spawns the go routine to monitor the certificate and ensure it is
+// rotated/renewed as necessary. The chan will indicate once the started
+// go routine has exited
+func (m *CertMonitor) Start(ctx context.Context) (<-chan struct{}, error) {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	if m.running || m.cancel != nil {
+		return nil, fmt.Errorf("the CertMonitor is already running")
+	}
+
+	// create the top level context to control the go
+	// routine executing the `run` method
+	ctx, cancel := context.WithCancel(ctx)
+
+	// create the channel to get cache update events through
+	// really we should only ever get 10 updates
+	m.cacheUpdates = make(chan cache.UpdateEvent, 10)
+
+	// setup the cache watches
+	cancelWatches, err := m.setupCacheWatches(ctx)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("error setting up cache watches: %w", err)
+	}
+
+	// start the token update notifier
+	m.tokenUpdates = m.tokens.Notify(token.TokenKindAgent)
+
+	// store the cancel funcs
+	m.cancel = cancel
+	m.cancelWatches = cancelWatches
+
+	m.running = true
+	exit := make(chan struct{})
+	go m.run(ctx, exit)
+
+	return exit, nil
+}
+
+// Stop manually stops the go routine spawned by Start and
+// returns whether the go routine was still running before
+// cancelling.
+//
+// Note that cancelling the context passed into Start will
+// also cause the go routine to stop
+func (m *CertMonitor) Stop() bool {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	if !m.running {
+		return false
+	}
+
+	if m.cancel != nil {
+		m.cancel()
+	}
+
+	return true
+}
+
+// IsRunning returns whether the go routine to perform certificate monitoring
+// is already running.
+func (m *CertMonitor) IsRunning() bool {
+	m.l.Lock()
+	defer m.l.Unlock()
+	return m.running
+}
+
+// setupCacheWatches will start both the roots and leaf cert watch with a new child
+// context and an up to date ACL token. The watches are started with a new child context
+// whose CancelFunc is also returned.
+func (m *CertMonitor) setupCacheWatches(ctx context.Context) (context.CancelFunc, error) {
+	notificationCtx, cancel := context.WithCancel(ctx)
+
+	// copy the request
+	rootsReq := m.rootsReq
+
+	err := m.cache.Notify(notificationCtx, cachetype.ConnectCARootName, &rootsReq, rootsWatchID, m.cacheUpdates)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	// copy the request
+	leafReq := m.leafReq
+	leafReq.Token = m.tokens.AgentToken()
+
+	err = m.cache.Notify(notificationCtx, cachetype.ConnectCALeafName, &leafReq, leafWatchID, m.cacheUpdates)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	return cancel, nil
+}
+
+// handleCacheEvent is used to handle event notifications from the cache for the roots
+// or leaf cert watches.
+func (m *CertMonitor) handleCacheEvent(u cache.UpdateEvent) error {
```

Comment:
We should handle the case where `u.Err != nil` somewhere in this function. The `if !ok` checks end up catching this case and erroring since `u.Result == nil`, but we probably want to put the error message into the log output.